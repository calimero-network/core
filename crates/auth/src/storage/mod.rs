use std::collections::HashMap;
use std::sync::Arc;

use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::config::StorageConfig;

pub mod key_manager;
pub mod models;
pub mod providers;
pub mod registry;

// Re-export storage implementations and key manager for backward compatibility
pub use key_manager::KeyManager;
pub use models::{prefixes, Key};
pub use providers::memory::MemoryStorage;

/// Storage error
#[derive(Debug, Error)]
pub enum StorageError {
    #[error("Item not found")]
    NotFound,

    #[error("Storage error: {0}")]
    StorageError(String),

    #[error("Serialization error: {0}")]
    SerializationError(String),

    #[error("Validation error: {0}")]
    ValidationError(String),
}

/// Storage trait
#[async_trait]
pub trait Storage: Send + Sync + 'static {
    /// Get a value from storage
    ///
    /// # Arguments
    ///
    /// * `key` - The key to get
    ///
    /// # Returns
    ///
    /// * `Result<Option<Vec<u8>>, StorageError>` - The value if found
    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, StorageError>;

    /// Set a value in storage
    ///
    /// # Arguments
    ///
    /// * `key` - The key to set
    /// * `value` - The value to set
    ///
    /// # Returns
    ///
    /// * `Result<(), StorageError>` - Success or error
    async fn set(&self, key: &str, value: &[u8]) -> Result<(), StorageError>;

    /// Delete a value from storage
    ///
    /// # Arguments
    ///
    /// * `key` - The key to delete
    ///
    /// # Returns
    ///
    /// * `Result<(), StorageError>` - Success or error
    async fn delete(&self, key: &str) -> Result<(), StorageError>;

    /// Check if a key exists in storage
    ///
    /// # Arguments
    ///
    /// * `key` - The key to check
    ///
    /// # Returns
    ///
    /// * `Result<bool, StorageError>` - Whether the key exists
    async fn exists(&self, key: &str) -> Result<bool, StorageError>;

    /// List keys with a prefix
    ///
    /// # Arguments
    ///
    /// * `prefix` - The prefix to list keys for
    ///
    /// # Returns
    ///
    /// * `Result<Vec<String>, StorageError>` - The keys
    async fn list_keys(&self, prefix: &str) -> Result<Vec<String>, StorageError>;

    /// Get multiple values from storage
    ///
    /// # Arguments
    ///
    /// * `keys` - The keys to get
    ///
    /// # Returns
    ///
    /// * `Result<HashMap<String, Vec<u8>>, StorageError>` - The values for keys that exist
    async fn get_batch(&self, keys: &[String]) -> Result<HashMap<String, Vec<u8>>, StorageError> {
        // Default implementation uses single get operations
        let mut result = HashMap::new();
        for key in keys {
            if let Some(value) = self.get(key).await? {
                result.insert(key.clone(), value);
            }
        }
        Ok(result)
    }

    /// Set multiple values in storage
    ///
    /// # Arguments
    ///
    /// * `values` - The key-value pairs to set
    ///
    /// # Returns
    ///
    /// * `Result<(), StorageError>` - Success or error
    async fn set_batch(&self, values: &HashMap<String, Vec<u8>>) -> Result<(), StorageError> {
        // Default implementation uses single set operations
        for (key, value) in values {
            self.set(key, value).await?;
        }
        Ok(())
    }

    /// Delete multiple values from storage
    ///
    /// # Arguments
    ///
    /// * `keys` - The keys to delete
    ///
    /// # Returns
    ///
    /// * `Result<(), StorageError>` - Success or error
    async fn delete_batch(&self, keys: &[String]) -> Result<(), StorageError> {
        // Default implementation uses single delete operations
        for key in keys {
            // Ignore "not found" errors
            let _ = self.delete(key).await;
        }
        Ok(())
    }

    /// Create a secondary index
    ///
    /// # Arguments
    ///
    /// * `index_name` - The name of the index
    /// * `key` - The primary key
    /// * `index_key` - The index key
    ///
    /// # Returns
    ///
    /// * `Result<(), StorageError>` - Success or error
    async fn create_index(
        &self,
        index_name: &str,
        key: &str,
        index_key: &str,
    ) -> Result<(), StorageError> {
        // Store a reference from the index key to the primary key
        let index_storage_key = format!("index:{}:{}:{}", index_name, index_key, key);
        self.set(&index_storage_key, &[]).await
    }

    /// Find keys by a secondary index
    ///
    /// # Arguments
    ///
    /// * `index_name` - The name of the index
    /// * `index_key` - The index key
    ///
    /// # Returns
    ///
    /// * `Result<Vec<String>, StorageError>` - The primary keys found
    async fn find_by_index(
        &self,
        index_name: &str,
        index_key: &str,
    ) -> Result<Vec<String>, StorageError> {
        let index_prefix = format!("index:{}:{}", index_name, index_key);
        let keys = self.list_keys(&index_prefix).await?;

        // Extract primary keys from index keys
        let primary_keys = keys
            .iter()
            .filter_map(|index_key| {
                let parts: Vec<&str> = index_key.split(':').collect();
                if parts.len() >= 4 {
                    Some(parts[3].to_string())
                } else {
                    None
                }
            })
            .collect();

        Ok(primary_keys)
    }

    /// Delete an index
    ///
    /// # Arguments
    ///
    /// * `index_name` - The name of the index
    /// * `key` - The primary key
    /// * `index_key` - The index key
    ///
    /// # Returns
    ///
    /// * `Result<(), StorageError>` - Success or error
    async fn delete_index(
        &self,
        index_name: &str,
        key: &str,
        index_key: &str,
    ) -> Result<(), StorageError> {
        let index_storage_key = format!("index:{}:{}:{}", index_name, index_key, key);
        self.delete(&index_storage_key).await
    }

    /// Get storage health status
    ///
    /// # Returns
    ///
    /// * `Result<serde_json::Value, StorageError>` - Health information
    async fn health_check(&self) -> Result<serde_json::Value, StorageError> {
        // Implement a basic health check by default
        let health_key = "_health_check_key";
        let health_value = b"ok";

        // Try to write and read a value
        self.set(health_key, health_value).await?;
        let read_result = self.get(health_key).await?;
        let read_ok = read_result.is_some() && read_result.unwrap() == health_value;

        // Clean up
        let _ = self.delete(health_key).await;

        Ok(serde_json::json!({
            "status": if read_ok { "healthy" } else { "unhealthy" },
            "read_write_test": read_ok,
        }))
    }
}

/// Create a storage instance based on the configuration
///
/// # Arguments
///
/// * `config` - The storage configuration
///
/// # Returns
///
/// * `Result<Arc<dyn Storage>, StorageError>` - The storage instance
pub async fn create_storage(config: &StorageConfig) -> Result<Arc<dyn Storage>, StorageError> {
    // Get all registered providers
    let providers = registry::get_all_providers();

    // Find a provider that supports this configuration
    for provider in providers {
        if provider.supports_config(config) {
            println!("provider: {}", provider.name());
            let storage = provider.create_storage(config)?;
            return Ok(storage);
        }
    }

    // If no registered provider is found, return an error
    Err(StorageError::StorageError(format!(
        "No registered storage provider found for configuration: {:?}",
        config
    )))
}

/// Helper function to serialize an object to JSON
///
/// # Arguments
///
/// * `value` - The value to serialize
///
/// # Returns
///
/// * `Result<Vec<u8>, StorageError>` - The serialized value
pub fn serialize<T: Serialize>(value: &T) -> Result<Vec<u8>, StorageError> {
    serde_json::to_vec(value).map_err(|e| StorageError::SerializationError(e.to_string()))
}

/// Helper function to deserialize an object from JSON
///
/// # Arguments
///
/// * `data` - The data to deserialize
///
/// # Returns
///
/// * `Result<T, StorageError>` - The deserialized value
pub fn deserialize<T: for<'de> Deserialize<'de>>(data: &[u8]) -> Result<T, StorageError> {
    serde_json::from_slice(data).map_err(|e| StorageError::SerializationError(e.to_string()))
}
