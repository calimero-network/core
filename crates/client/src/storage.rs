//! Token storage and management for Calimero client
//!
//! This module provides the core types and functionality for managing
//! JWT tokens used for API authentication.

// Standard library
use std::collections::HashMap;
use std::sync::Arc;

// External crates
use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;

/// JWT token pair for API authentication
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JwtToken {
    /// Access token for API requests
    pub access_token: String,
    /// Refresh token for renewing access tokens
    pub refresh_token: Option<String>,
    /// Token type (usually "Bearer")
    pub token_type: Option<String>,
    /// Expiration timestamp
    pub expires_at: Option<i64>,
    /// Additional token metadata
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl JwtToken {
    /// Create a new JWT token
    pub fn new(access_token: String) -> Self {
        Self {
            access_token,
            refresh_token: None,
            token_type: Some("Bearer".to_string()),
            expires_at: None,
            metadata: HashMap::new(),
        }
    }

    /// Create a new JWT token with refresh token
    pub fn with_refresh(access_token: String, refresh_token: String) -> Self {
        Self {
            access_token,
            refresh_token: Some(refresh_token),
            token_type: Some("Bearer".to_string()),
            expires_at: None,
            metadata: HashMap::new(),
        }
    }

    /// Check if the token is expired
    pub fn is_expired(&self) -> bool {
        if let Some(expires_at) = self.expires_at {
            let now = chrono::Utc::now().timestamp();
            now >= expires_at
        } else {
            false // No expiration set
        }
    }

    /// Check if the token will expire soon (within the given seconds)
    pub fn expires_soon(&self, within_seconds: i64) -> bool {
        if let Some(expires_at) = self.expires_at {
            let now = chrono::Utc::now().timestamp();
            let threshold = expires_at - within_seconds;
            now >= threshold
        } else {
            false
        }
    }

    /// Get the authorization header value
    pub fn auth_header(&self) -> String {
        let token_type = self.token_type.as_deref().unwrap_or("Bearer");
        format!("{} {}", token_type, self.access_token)
    }

    /// Add metadata to the token
    pub fn with_metadata(mut self, key: String, value: serde_json::Value) -> Self {
        self.metadata.insert(key, value);
        self
    }

    /// Get metadata value
    pub fn get_metadata(&self, key: &str) -> Option<&serde_json::Value> {
        self.metadata.get(key)
    }

    /// Check if the token has a refresh token
    pub fn has_refresh(&self) -> bool {
        self.refresh_token.is_some()
    }
}

impl Default for JwtToken {
    fn default() -> Self {
        Self {
            access_token: String::new(),
            refresh_token: None,
            token_type: Some("Bearer".to_string()),
            expires_at: None,
            metadata: HashMap::new(),
        }
    }
}

impl PartialEq for JwtToken {
    fn eq(&self, other: &Self) -> bool {
        self.access_token == other.access_token
    }
}

impl Eq for JwtToken {}

impl std::hash::Hash for JwtToken {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.access_token.hash(state);
    }
}

/// In-memory token cache for session management
#[derive(Debug, Clone)]
pub struct SessionTokenCache {
    tokens: Arc<RwLock<HashMap<String, JwtToken>>>,
}

impl SessionTokenCache {
    /// Create a new session token cache
    pub fn new() -> Self {
        Self {
            tokens: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Store tokens for a specific URL
    pub async fn store_tokens(&self, url: &str, tokens: &JwtToken) {
        let mut cache = self.tokens.write().await;
        cache.insert(url.to_string(), tokens.clone());
    }

    /// Get tokens for a specific URL
    pub async fn get_tokens(&self, url: &str) -> Option<JwtToken> {
        let cache = self.tokens.read().await;
        cache.get(url).cloned()
    }

    /// Remove tokens for a specific URL
    pub async fn remove_tokens(&self, url: &str) {
        let mut cache = self.tokens.write().await;
        cache.remove(url);
    }

    /// Clear all cached tokens
    pub async fn clear_all(&self) {
        let mut cache = self.tokens.write().await;
        cache.clear();
    }

    /// Check if tokens exist for a URL
    pub async fn has_tokens(&self, url: &str) -> bool {
        let cache = self.tokens.read().await;
        cache.contains_key(url)
    }

    /// Get all cached URLs
    pub async fn get_cached_urls(&self) -> Vec<String> {
        let cache = self.tokens.read().await;
        cache.keys().cloned().collect()
    }
}

impl Default for SessionTokenCache {
    fn default() -> Self {
        Self::new()
    }
}

/// Global session cache instance
static SESSION_CACHE: once_cell::sync::Lazy<SessionTokenCache> =
    once_cell::sync::Lazy::new(SessionTokenCache::new);

/// Get the global session cache instance
pub fn get_session_cache() -> SessionTokenCache {
    SESSION_CACHE.clone()
}

/// Token validation result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenValidation {
    /// Whether the token is valid
    pub is_valid: bool,
    /// Whether the token is expired
    pub is_expired: bool,
    /// Whether the token will expire soon
    pub expires_soon: bool,
    /// Time until expiration in seconds (negative if expired)
    pub expires_in: i64,
    /// Validation errors if any
    pub errors: Vec<String>,
}

impl TokenValidation {
    /// Create a validation result for a token
    pub fn new(token: &JwtToken) -> Self {
        let now = chrono::Utc::now().timestamp();
        let expires_in = token.expires_at.unwrap_or(0) - now;
        let is_expired = expires_in <= 0;
        let expires_soon = expires_in > 0 && expires_in <= 300; // 5 minutes

        let mut errors = Vec::new();
        if token.access_token.is_empty() {
            errors.push("Access token is empty".to_string());
        }
        if is_expired {
            errors.push("Token is expired".to_string());
        }

        Self {
            is_valid: errors.is_empty() && !is_expired,
            is_expired,
            expires_soon,
            expires_in,
            errors,
        }
    }

    /// Check if the token needs refresh
    pub fn needs_refresh(&self) -> bool {
        self.expires_soon || self.is_expired
    }
}
