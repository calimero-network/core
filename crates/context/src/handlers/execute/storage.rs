#![allow(clippy::mem_forget, reason = "ouroboros uses it")]

use core::cell::RefCell;
use core::mem;
use std::sync::Arc;

use calimero_primitives::context::ContextId;
use calimero_runtime::store::{Key, Storage, Value};
use calimero_store::layer::temporal::Temporal;
use calimero_store::layer::{ReadLayer, WriteLayer};
use calimero_store::{key, Store};
use ouroboros::self_referencing;

#[self_referencing]
pub struct ContextStorage {
    context_id: ContextId,
    store: Store,

    #[covariant]
    #[borrows(mut store)]
    inner: Temporal<'this, 'static, Store>,
    // todo! unideal, will revisit the shape of WriteLayer to own keys (since they are now fixed-sized)
    keys: RefCell<Vec<Arc<key::ContextState>>>,
}

// safety: ContextStorage is constructed exclusively for the runtime
//         which maintains exclusive access in a single-threaded environment
//         before returning the same instance back to the constructor
//         which then proceeds to directly commit any written data
//         never having multiple references to this same instance
//         --
//         we can eventually get rid of this when Slice<'_>: Send
//         ref: https://github.com/calimero-network/core/commit/455fe09ca9be09df17046584a3ef6cd28564e01a
unsafe impl Send for ContextStorage {}

impl ContextStorage {
    pub fn from(store: Store, context_id: ContextId) -> Self {
        ContextStorageBuilder {
            context_id,
            store,
            inner_builder: |store| Temporal::new(store),
            keys: RefCell::default(),
        }
        .build()
    }

    fn state_key(&self, key: &[u8]) -> Option<&'static key::ContextState> {
        let mut state_key = [0; 32];

        (key.len() <= state_key.len()).then_some(())?;

        state_key[..key.len()].copy_from_slice(key);

        let mut keys = self.borrow_keys().borrow_mut();

        let context_id = self.borrow_context_id();

        keys.push(Arc::new(key::ContextState::new(*context_id, state_key)));

        // safety: TemporalStore lives as long as Self, so the reference will hold
        //         plus, we never return a reference to the keys externally
        unsafe {
            mem::transmute::<Option<&key::ContextState>, Option<&'static key::ContextState>>(
                keys.last().map(|x| &**x),
            )
        }
    }

    pub fn commit(mut self) -> eyre::Result<Store> {
        self.with_inner_mut(|inner| inner.commit())?;

        Ok(self.into_heads().store)
    }

    pub fn is_empty(&self) -> bool {
        self.borrow_inner().is_empty()
    }
}

impl Storage for ContextStorage {
    fn get(&self, key: &Key) -> Option<Vec<u8>> {
        let key = self.state_key(key)?;

        let slice = self.borrow_inner().get(key).ok()??;

        Some(slice.into_boxed().into_vec())
    }

    fn remove(&mut self, key: &Key) -> Option<Vec<u8>> {
        let key = self.state_key(key)?;

        self.with_inner_mut(|inner| {
            let old = inner
                .get(key)
                .ok()
                .flatten()
                .map(|slice| slice.into_boxed().into_vec());

            inner.delete(key).ok()?;

            old
        })
    }

    fn set(&mut self, key: Key, value: Value) -> Option<Value> {
        let key = self.state_key(&key)?;

        self.with_inner_mut(|inner| {
            let old = inner
                .has(key)
                .ok()?
                .then(|| inner.get(key).ok().flatten())
                .flatten()
                .map(|slice| slice.into_boxed().into_vec());

            inner.put(key, value.into()).ok()?;

            old
        })
    }

    fn has(&self, key: &Key) -> bool {
        let Some(key) = self.state_key(key) else {
            return false;
        };

        self.borrow_inner().has(key).unwrap_or_default()
    }
}
