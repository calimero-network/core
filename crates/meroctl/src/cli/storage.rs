use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};

use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JwtToken {
    pub access_token: String,
    pub refresh_token: Option<String>,
}

/// Simple in-memory cache for external connection tokens
/// These tokens are only kept for the duration of the session
#[derive(Debug, Default)]
pub struct SessionTokenCache {
    tokens: Mutex<HashMap<String, JwtToken>>,
}

impl SessionTokenCache {
    pub fn new() -> Self {
        Self::default()
    }

    /// Store tokens for an external connection (session only)
    pub fn store_tokens(&self, url: &Url, tokens: &JwtToken) {
        let key = format!("external_{}", url.host_str().unwrap_or("unknown"));
        if let Ok(mut cache) = self.tokens.lock() {
            drop(cache.insert(key, tokens.clone()));
        }
    }

    /// Get tokens for an external connection
    pub fn get_tokens(&self, url: &Url) -> Option<JwtToken> {
        let key = format!("external_{}", url.host_str().unwrap_or("unknown"));
        self.tokens.lock().ok()?.get(&key).cloned()
    }

    /// Update tokens for an external connection
    pub fn update_tokens(&self, url: &Url, tokens: &JwtToken) {
        self.store_tokens(url, tokens);
    }

    /// Clear all cached tokens
    pub fn clear_all(&self) {
        if let Ok(mut cache) = self.tokens.lock() {
            cache.clear();
        }
    }
}

/// Global session cache instance
static SESSION_CACHE: OnceLock<Arc<SessionTokenCache>> = OnceLock::new();

/// Get the global session cache instance
pub fn get_session_cache() -> &'static Arc<SessionTokenCache> {
    SESSION_CACHE.get_or_init(|| Arc::new(SessionTokenCache::new()))
}
