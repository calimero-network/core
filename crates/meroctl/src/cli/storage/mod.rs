use async_trait::async_trait;
use eyre::Result as EyreResult;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use url::Url;

mod file;
mod keychain;
mod memory;

pub use file::FileStorage;
pub use keychain::KeychainStorage;
pub use memory::MemoryStorage;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JwtToken {
    pub access_token: String,
    pub refresh_token: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProfileConfig {
    pub node_url: Url,
    pub token: Option<JwtToken>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProfileTokens {
    pub profiles: HashMap<String, ProfileConfig>,
    pub current_profile: String,
}

#[async_trait]
pub trait TokenStorage: Send + Sync {
    async fn store_profile(&self, profile: &str, config: &ProfileConfig) -> EyreResult<()>;
    async fn load_profile(&self, profile: &str) -> EyreResult<Option<ProfileConfig>>;
    async fn remove_profile(&self, profile: &str) -> EyreResult<()>;
    async fn clear_all(&self) -> EyreResult<()>;
    async fn set_current_profile(&self, profile: &str) -> EyreResult<()>;
    // async fn get_current_profile(&self) -> EyreResult<Option<String>>;
    
    async fn list_profiles(&self) -> EyreResult<(Vec<String>, Option<String>)>;
    
    /// Get the current active profile and its config in one call to avoid multiple storage accesses
    async fn get_current_profile(&self) -> EyreResult<Option<(String, ProfileConfig)>>;
}

/// Create the appropriate storage backend based on system capabilities
pub fn create_storage() -> Box<dyn TokenStorage> {
    // Priority: Keychain -> File -> Memory (fallback)
    if KeychainStorage::is_available() {
        Box::new(KeychainStorage::new())
    } else {
        Box::new(FileStorage::new())
    }
}

/// Create in-memory storage for testing
pub fn create_memory_storage() -> Box<dyn TokenStorage> {
    Box::new(MemoryStorage::new())
} 