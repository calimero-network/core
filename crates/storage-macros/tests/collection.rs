#![allow(unused_crate_dependencies, reason = "Creates a lot of noise")]
//	Lints specifically disabled for integration tests
#![allow(
    non_snake_case,
    unreachable_pub,
    clippy::cast_lossless,
    clippy::cast_precision_loss,
    clippy::cognitive_complexity,
    clippy::default_numeric_fallback,
    clippy::exhaustive_enums,
    clippy::exhaustive_structs,
    clippy::expect_used,
    clippy::indexing_slicing,
    clippy::let_underscore_must_use,
    clippy::let_underscore_untyped,
    clippy::missing_assert_message,
    clippy::missing_panics_doc,
    clippy::mod_module_files,
    clippy::must_use_candidate,
    clippy::panic,
    clippy::print_stdout,
    clippy::tests_outside_test_module,
    clippy::unwrap_in_result,
    clippy::unwrap_used,
    reason = "Not useful in tests"
)]

use borsh::{to_vec, BorshDeserialize};
use calimero_storage::address::{Id, Path};
use calimero_storage::entities::{Data, Element};
use calimero_storage::exports::{Digest, Sha256};
use calimero_storage::interface::Interface;
use calimero_storage_macros::{AtomicUnit, Collection};
use calimero_test_utils::storage::create_test_store;

#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
struct Child {
    content: String,
    #[storage]
    storage: Element,
}

impl Child {
    fn new(path: &Path) -> Self {
        Self {
            content: String::new(),
            storage: Element::new(path),
        }
    }
}

#[derive(Collection, Clone, Debug, Eq, PartialEq, PartialOrd)]
#[children(Child)]
struct Group {
    #[child_ids]
    child_ids: Vec<Id>,
}

impl Group {
    fn new() -> Self {
        Self {
            child_ids: Vec::new(),
        }
    }
}

#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
struct Parent {
    title: String,
    #[collection]
    children: Group,
    #[storage]
    storage: Element,
}

impl Parent {
    fn new(path: &Path) -> Self {
        Self {
            title: String::new(),
            children: Group::new(),
            storage: Element::new(path),
        }
    }
}

#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
struct Simple {
    name: String,
    value: i32,
    #[storage]
    storage: Element,
}

impl Simple {
    fn new(path: &Path) -> Self {
        Self {
            name: String::new(),
            value: 0,
            storage: Element::new(path),
        }
    }
}

#[cfg(test)]
mod hierarchy {
    use super::*;

    #[test]
    fn parent_child() {
        let parent_path = Path::new("::root::node").unwrap();
        let mut parent = Parent::new(&parent_path);
        _ = parent.set_title("Parent Title".to_owned());

        let child_path = Path::new("::root::node::leaf").unwrap();
        let mut child = Child::new(&child_path);
        _ = child.set_content("Child Content".to_owned());

        assert_eq!(parent.title(), "Parent Title");

        // TODO: Add in tests for loading and checking children
    }

    #[test]
    fn compile_fail() {
        trybuild::TestCases::new().compile_fail("tests/compile_fail/collection.rs");
    }
}

#[cfg(test)]
mod hashing {
    use super::*;

    #[test]
    fn calculate_full_merkle_hash__cached_values() {
        let (db, _dir) = create_test_store();
        let interface = Interface::new(db);
        let mut parent = Parent::new(&Path::new("::root::node").unwrap());
        _ = parent.set_title("Parent Title".to_owned());
        assert!(interface.save(parent.id(), &mut parent).unwrap());
        assert_eq!(interface.children_of(&parent.children).unwrap(), vec![]);

        let mut child1 = Child::new(&Path::new("::root::node::leaf1").unwrap());
        let mut child2 = Child::new(&Path::new("::root::node::leaf2").unwrap());
        let mut child3 = Child::new(&Path::new("::root::node::leaf3").unwrap());
        _ = child1.set_content("Child 1 Content".to_owned());
        _ = child2.set_content("Child 2 Content".to_owned());
        _ = child3.set_content("Child 3 Content".to_owned());
        assert!(interface.save(child1.id(), &mut child1).unwrap());
        assert!(interface.save(child2.id(), &mut child2).unwrap());
        assert!(interface.save(child3.id(), &mut child3).unwrap());
        parent.children.child_ids = vec![child1.id(), child2.id(), child3.id()];
        assert!(interface.save(parent.id(), &mut parent).unwrap());

        let mut hasher0 = Sha256::new();
        hasher0.update(parent.id().as_bytes());
        hasher0.update(&to_vec(&parent.title).unwrap());
        hasher0.update(&to_vec(&parent.element().metadata()).unwrap());
        let expected_hash0: [u8; 32] = hasher0.finalize().into();

        let mut hasher1 = Sha256::new();
        hasher1.update(child1.id().as_bytes());
        hasher1.update(&to_vec(&child1.content).unwrap());
        hasher1.update(&to_vec(&child1.element().metadata()).unwrap());
        let expected_hash1: [u8; 32] = hasher1.finalize().into();
        let mut hasher1b = Sha256::new();
        hasher1b.update(expected_hash1);
        let expected_hash1b: [u8; 32] = hasher1b.finalize().into();

        let mut hasher2 = Sha256::new();
        hasher2.update(child2.id().as_bytes());
        hasher2.update(&to_vec(&child2.content).unwrap());
        hasher2.update(&to_vec(&child2.element().metadata()).unwrap());
        let expected_hash2: [u8; 32] = hasher2.finalize().into();
        let mut hasher2b = Sha256::new();
        hasher2b.update(expected_hash2);
        let expected_hash2b: [u8; 32] = hasher2b.finalize().into();

        let mut hasher3 = Sha256::new();
        hasher3.update(child3.id().as_bytes());
        hasher3.update(&to_vec(&child3.content).unwrap());
        hasher3.update(&to_vec(&child3.element().metadata()).unwrap());
        let expected_hash3: [u8; 32] = hasher3.finalize().into();
        let mut hasher3b = Sha256::new();
        hasher3b.update(expected_hash3);
        let expected_hash3b: [u8; 32] = hasher3b.finalize().into();

        let mut hasher = Sha256::new();
        hasher.update(&expected_hash0);
        hasher.update(&expected_hash1b);
        hasher.update(&expected_hash2b);
        hasher.update(&expected_hash3b);
        let expected_hash: [u8; 32] = hasher.finalize().into();

        assert_eq!(parent.calculate_merkle_hash().unwrap(), expected_hash0);
        assert_eq!(
            child1
                .calculate_full_merkle_hash(&interface, false)
                .unwrap(),
            expected_hash1b
        );
        assert_eq!(
            child2
                .calculate_full_merkle_hash(&interface, false)
                .unwrap(),
            expected_hash2b
        );
        assert_eq!(
            child3
                .calculate_full_merkle_hash(&interface, false)
                .unwrap(),
            expected_hash3b
        );
        assert_eq!(
            parent
                .calculate_full_merkle_hash(&interface, false)
                .unwrap(),
            expected_hash
        );
    }

    #[test]
    #[ignore]
    fn calculate_full_merkle_hash__recalculated_values() {
        // TODO: Later, tests should be added for recalculating the hashes, and
        // TODO: especially checking when the data has been interfered with or
        // TODO: otherwise arrived at an invalid state.
        todo!()
    }

    #[test]
    fn calculate_merkle_hash__child() {
        let mut child = Child::new(&Path::new("::root::node::leaf").unwrap());
        _ = child.set_content("Child Content".to_owned());

        let mut hasher = Sha256::new();
        hasher.update(child.id().as_bytes());
        hasher.update(&to_vec(&child.content).unwrap());
        hasher.update(&to_vec(&child.element().metadata()).unwrap());
        let expected_hash: [u8; 32] = hasher.finalize().into();

        assert_eq!(child.calculate_merkle_hash().unwrap(), expected_hash);
    }

    #[test]
    fn calculate_merkle_hash__parent() {
        let mut parent = Parent::new(&Path::new("::root::node").unwrap());
        _ = parent.set_title("Parent Title".to_owned());

        let mut hasher = Sha256::new();
        hasher.update(parent.id().as_bytes());
        hasher.update(&to_vec(&parent.title).unwrap());
        hasher.update(&to_vec(&parent.element().metadata()).unwrap());
        let expected_hash: [u8; 32] = hasher.finalize().into();

        assert_eq!(parent.calculate_merkle_hash().unwrap(), expected_hash);
    }
}

#[cfg(test)]
mod traits {
    use super::*;

    #[test]
    fn serialization_and_deserialization() {
        let path = Path::new("::root::node").unwrap();
        let mut unit = Simple::new(&path);

        _ = unit.set_name("Test Name".to_owned());
        _ = unit.set_value(42);

        let serialized = to_vec(&unit).unwrap();
        let deserialized = Simple::try_from_slice(&serialized).unwrap();

        assert_eq!(unit, deserialized);
        assert_eq!(unit.id(), deserialized.id());
        assert_eq!(unit.name(), deserialized.name());
        assert_eq!(unit.path(), deserialized.path());
        assert_eq!(unit.value(), deserialized.value());
        assert_eq!(unit.element().id(), deserialized.element().id());
        assert_eq!(unit.element().path(), deserialized.element().path());
        assert_eq!(unit.element().metadata(), deserialized.element().metadata());
    }
}
