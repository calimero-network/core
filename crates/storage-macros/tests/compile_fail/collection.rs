use calimero_storage::address::Path;
use calimero_storage::entities::{ChildInfo, Element};
use calimero_storage::interface::Interface;
use calimero_storage_macros::{AtomicUnit, Collection};
use calimero_test_utils::storage::create_test_store;

#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
struct Child {
    #[storage]
    storage: Element,
}

#[derive(Collection, Clone, Debug, Eq, PartialEq, PartialOrd)]
#[children(Child)]
struct Group {
    #[child_info]
    child_info: Vec<ChildInfo>,
}

#[derive(AtomicUnit, Clone, Debug, Eq, PartialEq, PartialOrd)]
struct Parent {
    group: Group,
	#[storage]
	storage: Element,
}

fn main() {
    fn child_type_specification() {
        let (db, _dir) = create_test_store();
        let interface = Interface::new(db);
        let parent: Parent = Parent {
            group: Group { child_info: vec![] },
            storage: Element::new(&Path::new("::root::node").unwrap()),
        };
        let _: Vec<Child> = interface.children_of(&parent.group).unwrap();

        // This should fail to compile if the child type is incorrect
        let _: Vec<Parent> = interface.children_of(&parent.group).unwrap();
    }
}
