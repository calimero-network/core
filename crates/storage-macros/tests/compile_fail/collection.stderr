error[E0308]: mismatched types
  --> tests/compile_fail/collection.rs:38:30
   |
38 |         let _: Vec<Parent> = interface.children_of(&parent.group).unwrap();
   |                -----------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Vec<Parent>`, found `Vec<Child>`
   |                |
   |                expected due to this
   |
   = note: expected struct `Vec<Parent>`
              found struct `Vec<Child>`
